### PR TITLE
Reintroduce capability conflict bugfixes

### PR DESCRIPTION
**Pull request by jvandort**
Reintroduce https://github.com/gradle/gradle/pull/29853

### Reviewing cheatsheet

- [ ] Add proper test to verify the behavior that this fixes

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java [hunk 1]
```diff
@@ -95,65 +100,59 @@ public boolean conflictExists() {
...
     public boolean hasConflicts() {
         return !conflicts.isEmpty();
     }
 
     @Override
     public void resolveNextConflict(Action<ConflictResolutionResult> resolutionAction) {
-        CapabilityConflict conflict = conflicts.poll();
+        String capabilityInConflict = conflicts.remove();
+        CapabilityConflict conflict = capabilityIdToConflict.remove(capabilityInConflict).withSelectedNodes();
+        capabilityWithoutVersionToNodes.put(capabilityInConflict, conflict.nodes);
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java:118]: I wonder if we still need this line. now that we are filtering in `registerCandidate`

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java (hunk 1 above)
Comment (ljacomet) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java:118]: I believe we do. While we filter at registration time, a node could be deselected between the registration and the conflict resolution itself.

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
Can confirm `sanityCheck` on the gradle/gradle build passes with these changes. 
A failure when updating the gradle/gradle build to use these changes caused the original revert.

**Pull request merged**